src/azure-cli-core/azure/cli/core/auth/msal_authentication.py

```diff
@@ -138,3 +138,27 @@ def get_token(self, *scopes, **kwargs):
         result = self.acquire_token_for_client(scopes, **kwargs)
         check_result(result)
         return build_sdk_access_token(result)
+
+
+class CloudShellCredential:  # pylint: disable=too-few-public-methods
+
+    def __init__(self):
+        from .identity import AZURE_CLI_CLIENT_ID
+        self._msal_app = PublicClientApplication(
+            AZURE_CLI_CLIENT_ID,  # Use a real client_id, so that cache would work
+            # TODO: This PoC does not currently maintain a token cache;
+            #   Ideally we should reuse the real MSAL app object which has cache configured.
+            # token_cache=...,
+        )
```

Comment:
The current `msrestazure`-based implementation uses no token cache. We can add token cache in the future.